src/transformers/modeling_utils.py

```diff
@@ -1049,15 +1156,18 @@ def _generate_beam_search(
                 next_sent_beam = []
 
                 # next tokens for this sentence
-                for idx, score in zip(next_tokens[batch_idx], next_scores[batch_idx]):
-
+                for i, (idx, score) in enumerate(zip(next_tokens[batch_idx], next_scores[batch_idx])):
                     # get beam and word IDs
                     beam_id = idx // vocab_size
                     token_id = idx % vocab_size
 
                     effective_beam_id = batch_idx * num_beams + beam_id
+
                     # add to generated hypotheses if end of sentence
-                    if eos_token_ids is not None and token_id.item() in eos_token_ids:
+                    if (eos_token_ids is not None) and (token_id.item() in eos_token_ids):
+                        # when passed to num_beams hypotheses, continue
+                        if i >= num_beams:
```

Comment:
This if-continue is also needed to pass the Integration tests - also seem quite hacky to me. Did not really see the logic behind it @sshleifer

Reply:
It confused me big time!

**What's going on?**
Suppose we have 4 beams. 
We took the top 8 possible next words for each of them.

If we see a token_id that's not EOS, we `next_sent_beam.append`  the associated `(score, token_id, beam_id)`
If in the top 4 of those words we see EOS, we finalize a hypothesis.
If we are outside the top 4 and we see EOS, we don't want to finalize a hypothesis, or continue with EOS in next_words. Observe that we are guaranteed to have already seen a better word for this beam.

If you don't have the logic, you finish a little earlier (you finalize more quickly) and get somewhat lower rouge. Also this is not just Bart, but all of fairseq.

Reply:
Got it. This makes sense to me then - will add a better comment and change variable name to make it beter understandable. I think as well that this is then not related to Bart but holds true in general.